File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    longest_dim = 0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        def dfs(node):
            if not node:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)
            self.longest_dim = max(left+right, self.longest_dim)
            return 1 + max(left, right)

        dfs(root)
        return self.longest_dim

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int res;
    void dfs(TreeNode* node, int greatest) {
        if (node == nullptr) return;
        if (node->val >= greatest) res+=1;
        if (node->left) dfs(node->left, max(node->val,greatest));
        if (node->right) dfs(node->right, max(node->val,greatest));
    }

    int goodNodes(TreeNode* root) {
        res = 0;
        dfs(root,root->val);
        return res;
    }
};
